FIX: Allow different colormap name from registered name

greglucas opened the pull request:
## PR Summary

When registering a colormap you can use a different name than the `cmap.name` attribute now. This will set the colormap based on the registered name rather than `cmap.name`.

closes #5087

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [-] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [-] New plotting related features are documented with examples.

**Release Notes**
- [-] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [-] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [-] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented on lib/matplotlib/pyplot.py:
```diff
@@ -2180,9 +2180,13 @@ def set_cmap(cmap):
+    # The name of the colormap may be different than the registered name.
+    # We want the registered name for the rc setting
+    name = getattr(cmap, "name", cmap)
+    # Run this through get_cmap to confirm this is a valid cmap before
```
Can this be clearer which "this" is meant here?  I almost commented based on this comment that the next line should be `cmap = get_cmap(name)`, but am no longer sure which was intended.

greglucas replied:
I _think_ we want to use `get_cmap(cmap)` here in case someone didn't register the colormap? Although that would also mean the rc paramter wouldn't work then either...

jklymak commented on lib/matplotlib/pyplot.py:
```diff
@@ -2180,9 +2180,13 @@ def set_cmap(cmap):
+    name = getattr(cmap, "name", cmap)
+    # Run the cmap (str or Colormap) through get_cmap to validate
+    # the cmap before updating the rc parameters
     cmap = get_cmap(cmap)
```
Perhaps clearer to return a nonesense variable here?  Even with he comments, I thought the next line was a typo...

```suggestion
    _ = get_cmap(cmap)
```

ksunden commented on lib/matplotlib/pyplot.py:
```diff
@@ -2180,9 +2180,13 @@ def set_cmap(cmap):
     cmap = get_cmap(cmap)
-
-    rc('image', cmap=cmap.name)
+    rc('image', cmap=name)
```
If I do:

```python
my_cmap = mpl.colormaps["viridis"].copy()
my_cmap.set_bad("FF00FF")
mpl.colormaps.register("my-cmap", my_cmap) # note, did not update cmap.name
plt.set_cmap(my_cmap)  # note, passing instance, not name
```

Won't this get `"viridis"` from the `getattr`, and thus set that as the rc cmap? (It will apply the "correct" cmap to `gci`, but not to future mappables)

Not sure how to square that without a) doing a reverse lookup in the registry or b) requiring that the names match when registered.

Perhaps this change is still better than no change, but not sure it's without its rough edges still... certainly it makes it possible to use a mismatched cmap, but you do need to get it by the registry name, not passing the cmap object.

greglucas commented:
Of course, a bug that has been hanging around for a while isn't _super_ easy to fix.

@ksunden, good catch and explanation about that edge-case. You said it much better than me in one of the other comments:
> I think we want to use get_cmap(cmap) here in case someone didn't register the colormap? Although that would also mean the rc paramter wouldn't work then either...

Question: Is it OK to assign a colormap object to the `image.cmap` rcParam? I think that is the only way we can do this because otherwise, we get what Kyle was mentioning above with the mismatched registered vs actual cmap name.

I just pushed up a commit with that, which I think is the more robust way to do this, but all other rcParams I think are strings, so I'm not sure if this is OK during runtime.

greglucas commented on lib/matplotlib/pyplot.py:
```diff
@@ -2180,9 +2180,10 @@ def set_cmap(cmap):
     cmap = get_cmap(cmap)
-
-    rc('image', cmap=cmap.name)
+    rc('image', cmap=cmap)
```
See here for changing the string name to the object itself.

tacaswell commented:
I was about to say I was wary of just putting the colormap object in rcparams....but the validator allows it from bb2de5e02d89a70cec5854154ae1f2b03419912a so we support that from 3.4 on.

Maybe the logic should be if the named lookup is itself then use the string otherwise the object?

greglucas commented:
> Maybe the logic should be if the named lookup is itself then use the string otherwise the object?

I think we'd need to add a bit of unsightly code and try/except for this case because the name of the cmap may not be in the lookup table if it was registered under a different name, thus throwing a ValueError cmap is not a valid name.

It seems like quite a bit of complication to add and I'm not sure how much the string nature matters here since we are only doing this during runtime? Unless someone is exporting the rcParams to a file and hoping to reimport them later with a specific string...

greglucas commented:
Since I changed this to updating the rcParams with an object instead of string, I've remilestoned to 3.8 instead because some people may be relying on rcParams of a certain type and it doesn't seem quite right for a patch release and this bug has been around for a while already. Happy to change back if people think otherwise.

timhoffm reviewed (changes requested):
I'm 👎 on registering a colormap with a name different from its original name. This is confusing up to dangerous.

By default `colormaps.register(cmap)` uses `cmap.name`. The only reason to use `register(cmap, name=other_name)` is that the user intentionally wants to register the colormap under a different name. We can interpret this as authorization to rename the colormap itself as well.

Note that `ColormapRegistry.register()` stores a copy of the given colormap to avoid any later modifcations. Since renaming would only apply to this copy, it cannot have unintended side-effects.

greglucas commented:
@timhoffm, thank you for taking a look at this and coming up with that suggestion! I like the idea of setting the name on the object that gets registered to keep it consistent and I think that flows nicely with expectations from an end-user perspective (I forgot we make a copy upon registering now). I've updated the code accordingly. Note that I also removed `cmap1.name == cmap2.name` in the `__eq__` method, which really shouldn't matter from a comparison perspective, we should only bother with the lookup tables which are what visually identify the colormaps being equal.

timhoffm commented on lib/matplotlib/tests/test_colors.py:
```diff
@@ -195,10 +195,6 @@ def test_colormap_equals():
     # Make sure we can compare different sizes without failure
     cm_copy._lut = cm_copy._lut[:10, :]
     assert cm_copy != cmap
-    # Test different names are not equal
-    cm_copy = cmap.copy()
-    cm_copy.name = "Test"
-    assert cm_copy != cmap
```
I suggest to keep this and test that colormaps with different names are still equal.

timhoffm reviewed (approved):
Modulo, restoring the equal test.
